### PR TITLE
Binary threshold on saturation instead of luminescence value.

### PR DESCRIPTION
**Pull request by Franz1960**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### CODE REVIEW

**Review by antonfirsov — CHANGES_REQUESTED** [review]
This is gonna be a breaking change in the proposed form. (New overload with a default arg replacing the existing one.) You should refactor default args. to separate overloads.

You also need to add tests in form of a new `[Theory]` to the following test class:
https://github.com/SixLabors/ImageSharp/blob/a1784a67bf492b1b02e99655a39aa5e5acf81988/tests/ImageSharp.Tests/Processing/Processors/Binarization/BinaryThresholdTest.cs

See second part of https://github.com/SixLabors/ImageSharp/issues/1498#issuecomment-758564510 for guidance on extending tests. (Difference is that the recommendations there for `ResizeTests`, replace them with `BinaryThresholdTest` when reading.)

Summary: 3 things to do in the PR:
1. Enum instead of `bool` arg
2. Separate overloads
3. Add tests

File: src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs
Comment (Franz1960) [reply src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs:23]: Thank you for the recommendations. I should find time til tomorrow.

**Comment by Franz1960** [comment]
@antonfirsov: Is the overload in the current form still a breaking change?

**Comment by antonfirsov** [comment]
@Franz1960 yeah, the problem is that at IL level `Foo(a,b=0)` is not the same as `Foo(a)` + a separate `Foo(a,b)`. Here is a short explanation with IL:
https://stackoverflow.com/a/12671182

And here is a (somewhat) longer article:
https://haacked.com/archive/2010/08/10/versioning-issues-with-optional-arguments.aspx/

TLDR: If we were designing this API from scratch, we could just use default args, but since it's not the case, we need to add separate overloads for each version in order to avoid unexpected `MissingMethodException`-s for users who already took dependency on any of the overloads, and are updating ImageSharp *without rebuilding their app*.

File: src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs
Comment (antonfirsov) [inline src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs:37]: Additionally, `Rectangle` should be the last argument when there `treshold` and/or `valueType`  is present (for consistency).

File: src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs:11]: `BinaryThresholdComponentKind` ? Naming is hard...

Comment (Franz1960) [reply src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs:11]: Oh yes, that's true.
My naming candidates have been
- BinaryThresholdColorComponent
- BinaryThresholdValue
- BinaryThresholdValueType
Now Jim added
- BinaryThresholdComponentKind
My intention would be to make the enumeration expandable for each color component of each color model.
I'm open regarding names. How can we find the best one?

File: src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs
Comment (Franz1960) [reply src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs:37]: Ok, thank you for clarification. I will implement it soon.

One more question: Code coverage test continues to fail even though I have extended the test methods to pass the branch with saturation. What am I doing wrong?

File: src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs
Comment (brianpopow) [reply src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs:11]: What about `BinaryThresholdMethod`?

Comment (antonfirsov) [reply src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs:11]: If I get it right, it's not a method, it's just just a value selector.

File: src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs
Comment (antonfirsov) [reply src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs:37]: I think the problem is on the codecov (tool) side not on yours. I'm not an expert on that, @JimBobSquarePants any idea what's going on?

Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs:37]: What’s happening with that new test? We’re doing a new comparison with saturation but I do not see how since the submodule has not been updated.

Comment (antonfirsov) [reply src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs:37]: There was no validation in the old tests, so there is none in the new one (just debug save).

Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs:37]: https://github.com/SixLabors/ImageSharp/pull/1540/files#diff-88fae6a5fdda36fa3f6f2ba33c87415448b8db16c5320bdf437ee75235381be1R83

Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs:37]: I’m on my phone so can’t check what that verify method is doing

Comment (antonfirsov) [reply src/ImageSharp/Processing/Extensions/Binarization/BinaryThresholdExtensions.cs:37]: If memory serves well, that method just checks if the area outside the rect was left unchanged.

File: src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs
Comment (antonfirsov) [reply src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs:11]: `BinaryThresholdComponentKind` feels better than `BinaryThresholdValueType`, but based on the semantics you described, my favorite from the list above is `BinaryThresholdColorComponent`. I don't think we need a postfix like `*Type` or `*Kind` when the name already ends to `*Component`. @JimBobSquarePants thoughts?

File: tests/ImageSharp.Tests/Processing/Processors/Binarization/BinaryThresholdTest.cs
Comment (antonfirsov) [inline tests/ImageSharp.Tests/Processing/Processors/Binarization/BinaryThresholdTest.cs:66]: @Franz1960 can you please double check if the test results saved to your local `tests/Images/ActualOutput/BinaryThresholdTest/` meet your expectations (Saturation vs Luminance)?

Normally we'd have a validation call here in these functional tests (`image.CompareToReferenceOutput(...)`), comparing the result to a pre-existing reference output, but unfortunately `BinaryThresholdTest` haven't been updated to have this validation.

I don't want to block this PR on this, but my testing OCD tells that we should at least visually validate the output.

File: src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs:11]: Yeah... `BinaryThresholdComponent` works well. We've managed to avoid the `**Kind` suffix elsewhere.

File: tests/ImageSharp.Tests/Processing/Processors/Binarization/BinaryThresholdTest.cs
Comment (Franz1960) [reply tests/ImageSharp.Tests/Processing/Processors/Binarization/BinaryThresholdTest.cs:71]: I included comparisons to ref outputs. Used a rgb image which is better fitted to saturation operations. Added a PR in 
Imagesharp.Tests.Images repo with new ReferenceOutput images. Is this the right way to bring the ref images to work?

Comment (JimBobSquarePants) [reply tests/ImageSharp.Tests/Processing/Processors/Binarization/BinaryThresholdTest.cs:71]: Yep, now use 

``` bash
git submodule foreach git pull origin master
```
In your branch root to pull those reference images in. 

I'm going to pull down your fork later and do a performance pass as I've spotted scope for some improvement.

File: src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs
Comment (JimBobSquarePants) [inline src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs:27]: This isn't part of the original specification.

Comment (Franz1960) [reply src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs:27]: Ok, I see your point. I would see a certain degree of general purpose because saturation is really a problem at the dark end of the color space. At the light end saturation is always low, but rgb(0,1,0) has high saturation though it's practically black. The proposed Colorfulness_L10 works well in our toad application, and different L thresholds might be realised. A generalization with L threshold as argument would be possible but perhaps less convenient for users.

I fully understand if you don't want to have this in your lib, but we do need it. Would you recommend a fork? or is there a way to make extensions like that without integrating it?

Thanks and cheers,
Franz

Comment (antonfirsov) [reply src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs]: I don't like the idea of introducing enum-driven arbitrary strategies (they may be useless for most users), but we can generalize the feature by passing something like a `Func<Rgba32,byte,bool>` or `Func<Vector4,float,bool>` or equivalent to the processor. It's not gonna be trivial work, but I'm happy to guide you with such a contribution if you're interested.

Alternatively, you can just copy the code of `BinaryThresholdProcessor` and `BinaryThresholdProcessor<T>` into your application, and apply the changes there. Forking the entire project is unnecessary.

Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs]: Apologies for the silence. Day job is eating up all my waking hours. 

I think even a func overload would have to be very carefully designed so it wasn’t a performance killer and feels very niche. I’m happy with the general enum but specific cases can be delivered like Anton said. The library is designed to be extensible.

Comment (Franz1960) [reply src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs]: First I want to thank you for your time spent on this PR, both of you.

My arbitrarily defining a 'Colorfulness_L10' color component was too pragmatic. I will have a closer look on color standards like BT.709 or JPEG YCbCr. We need a scalar value to compare to a threshold. This value has to represent colorfulness or chromaticity to the human eye for my purposes. The HSL saturation value comes near but fails for dark colors. The distance of Cb or Cr to the achromatic level AL would be a candidate.

    Chroma = Max(Abs(Cb - AL), Abs(Cr - AL))

Would that be general purpose in your eyes?

Comment (antonfirsov) [reply src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs]: I like the approach but my understanding of colorspaces is not good enough not to evaluate the proposal.

Quick sanity test: How would you name this "mode"? How would you explain it in docs? Have you tested it already?

Comment (Franz1960) [reply src/ImageSharp/Processing/Processors/Binarization/BinaryThresholdProcessor.cs]: I think I understand the software requirements better now and will investigate the issue more thoroughly first. When I have a suitable solution from my point of view, I will make a corresponding proposal. Until then I will work on a branch without PRs. Maybe I will do a merge to saturation_threshold from time to time to see the results of the tests. Please just ignore and let the PR stay open! I will get back to you when I have a state that is ok from my point of view.

**Comment by antonfirsov** [comment]
Marking this as a draft while discussing the new proposal.

@Franz1960 thanks for your interest and patience, I really appreciate it!

**Comment by Franz1960** [comment]
If my PR #17 on Imagesharp.Tests.Images will be merged, these tests should run through.

**Comment by JimBobSquarePants** [comment]
Test images merged. I'll do that performance pass later.

**Comment by antonfirsov** [comment]
@Franz1960 do the new tests pass locally for you? I see test failures both because of missing reference output and big difference in output.

SixLabors/Imagesharp.Tests.Images#17 should be most likely reverted and redone.

When doing so, I recommend to run the [optipng tool](https://github.com/SixLabors/Imagesharp.Tests.Images/tree/master/tools) on the new reference images, they seem to be unreasonably large resulting in >1MB addition to the repo.

**Comment by Franz1960** [comment]
@antonfirsov Locally all tests passed. The optipng tool only shrinked the small files a little (black white only) so I did not use it.
I think the merge of Imagesharp.Tests.Images PR #17 had not been complete (it had been originally rejected by @JimBobSquarePants), so I make a new branch and PR #18, which has to be merged before tests will pass.

**Comment by JimBobSquarePants** [comment]
@Franz1960 I didn't reject anything...

https://github.com/SixLabors/Imagesharp.Tests.Images/pull/17

Btw you keep linking to the wrong locations. Hashes only work within the same repository.

**Comment by Franz1960** [comment]
Sorry I obvously don't understand the workflow with the Imagesharp.Tests.Images submodule. I forked it (Franz1960/Imagesharp.Tests.Images), made a branch "saturation_threshold", pushed the new and changed test images and made a PR (#17). @JimBobSquarePants commented it was beyond the accepted spec and I made changes accordingly and made commits of test images. Now in SixLabors/Imagesharp.Tests.Images / master all images are present, but the PR tests on ImageSharp failed.

How would I do this correctly?

**Comment by antonfirsov** [comment]
@Franz1960 based on what you describe you did things conceptually right, but some mistake happened on the way. What I see is that there are only 4 images pushed to the PR, while you modified 22 images in SixLabors/Imagesharp.Tests.Images#17. Maybe you did not point to the right commit with your submodule update? I recommend to do a fresh clone of your fork to an empty directory, and running the tests there. That should give you some clues on what went wrong.

FYI: We realized that the submodule workflow is a huge burden, so @JimBobSquarePants is currently working on getting rid of it by merging the reference images into the repository + introducing [git LFS](https://git-lfs.github.com/) for images. It will be a radical step with history rewrite on master. So an alternative to the approach I suggested in the previous paragraph is to wait until that work is done.

The preferred solution also depends on @JimBobSquarePants 's ETA on nuking down the submodules.

**Comment by JimBobSquarePants** [comment]
Ok.... So introducing Git LFS requires rewriting Git history which means this PR's history is now out of sync with our repository and GitHub has now closed it. 

My advice @Franz1960 , for simplicities sake, is to create a fresh clone of our repository and add the required changes to a new PR. You will be able to add the images now directly to the references folder but first ensure you have [Git Large File Storage](https://docs.github.com/en/github/managing-large-files/installing-git-large-file-storage) installed.

**Pull request closed without merging**
